src/persistent-connections/content.md

```diff
@@ -0,0 +1,465 @@
+- kube-proxy iterates through all the IP addresses and creates iptables rules
+
+So you could ignore kube-proxy all together and always use the list of endpoints collected by the Headless Service to load balance requests client-side.
+
+_But can you imagine adding that logic to all apps deployed in the cluster?_
+
+If you have an existing fleet of applications, this might sound like an impossible task.
+
+But there's an alternative.
+
+## Service meshes to the rescue
+
+You probably already noticed that the client-side load balancing strategy is quite standard.
+
+When the app starts, it should
+
+- retrieve a list of IP addresses from the Service
+- open and maintain a pool of connections
+- periodically refresh the pool by adding and removing endpoints
+
+As soon as it wishes to make a request, it should:
+
+- pick one of the available connections using a predefined logic such as round-robin
+- issue the request
+
+The steps above are valid for WebSockets connections as well as gRPC and AMQP.
+
+You could extract that logic in a separate library and share it with all apps.
+
+Instead of writing a library from scratch, you could use a Service mesh such as Istio or Linkerd.
+
+Service meshes augment your app with a new process that:
+
+- automatically discovers IP addresses Services
+- inspects connections such as WebSockets and gRPC
+- load-balances requests using the right protocol
+
+Service meshes can help you to manage the traffic inside your cluster, but they aren't exactly lightweight.
+
+Other options include using a library such as Netflix Ribbon, a programmable proxy such as Envoy or just ignore it.
+
+_What happens if you ignore it?_
+
+You can ignore the load balancing and still don't notice any change.
+
+There are a couple of scenarios that you should consider.
+
+If you have more client than servers, there should be limited issues.
+
+Imagine you have five clients opening persistent connections to two servers.
+
+Even if there's no load balancing, both servers likely utilised.
+
+![More clients than servers](assets/many-clients-fewer-servers.svg)
+
+The connections might not be distributed evenly (perhaps four ended up connecting to the same server), but overall there's a good chance that both servers are utilised.
+
+What's more problematic is the opposite scenario.
+
+If you have a few clients and more servers, you might have some underutilised resources and a potential bottleneck.
+
+Imagine having two clients and five servers.
+
+At best, two persistent connections to two servers are opened.
+
+The remaining servers are not used at all.
+
+![More servers than clients](assets/fewer-servers-many-clients.svg)
+
+If the two servers can't handle the traffic generated by the clients, horizontal scaling won't help.
+
+## Summary
+
+Kubernetes Services are designed to cover most common uses for web applications.
+
+However, as soon as you start dealing with persistent connections such as a database or gRPC service, they fall apart.
```

Comment:
However, as soon as you start working with **application protocols that use persistent TCP connections**, such as databases, gRPC, or WebSockets, they fall apart.